pkg/experiment/metastore/blockcleaner/block_cleaner.go

```diff
@@ -0,0 +1,417 @@
+package blockcleaner
+
+import (
+	"context"
+	"encoding/binary"
+	"flag"
+	"fmt"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/thanos-io/objstore"
+	"go.etcd.io/bbolt"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/grafana/pyroscope/pkg/experiment/metastore/raftlogpb"
+	"github.com/grafana/pyroscope/pkg/util"
+)
+
+const (
+	removedBlocksBucketName = "removed-blocks"
+)
+
+var removedBlocksBucketNameBytes = []byte(removedBlocksBucketName)
+
+type Config struct {
+	CompactedBlocksCleanupInterval time.Duration `yaml:"compacted_blocks_cleanup_interval"`
+	CompactedBlocksCleanupDelay    time.Duration `yaml:"compacted_blocks_cleanup_delay"`
+}
+
+func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
+	f.DurationVar(&cfg.CompactedBlocksCleanupDelay, prefix+"compacted-blocks-cleanup-delay", time.Minute*30, "The grace period for permanently deleting compacted blocks.")
+	f.DurationVar(&cfg.CompactedBlocksCleanupInterval, prefix+"compacted-blocks-cleanup-interval", time.Minute, "The interval at which block cleanup is performed.")
+}
+
+type CleanerLifecycler interface {
+	Cleaner
+
+	Start()
+	Stop()
+}
+
+type Cleaner interface {
+	MarkBlock(shard uint32, tenant string, blockId string, deletedTs int64) error
+	IsMarked(blockId string) bool
+
+	RemoveExpiredBlocks(now int64) error
+}
+
+type RaftLog[Req, Resp proto.Message] interface {
+	ApplyCommand(req Req) (resp Resp, err error)
+}
+
+type RaftLogCleanBlocks RaftLog[*raftlogpb.CleanBlocksRequest, *anypb.Any]
```

Comment:
> We could benefit from delegating managing the raftlogpb.CleanBlocksRequest to a separate component

Agreed.

> // Note that tombstones is a widely used term: https://en.wikipedia.org/wiki/Tombstone_(data_store). // Here and below I use word "mark", but I'd use tombstone, if I were to implement this.

I know what a tombstone is and its usage. It is more nuanced than that because this component, for the better or worse, does more than managing tombstones. A large part of its responsibility is doing garbage collection which the word "tombstone" doesn't communicate. If it were to remain one component (which doesn't have to be the case), "cleaner" encapsulates the two responsibilities more accurately. 

> type DeletionMarkLoader interface {

This part is actually missing currently which will cause a few bugs, good catch!

> `DeletionMarkChecker`, `DeletionMarkAdder`, `BlockRemover`

These are part of a single interface because the "user" is essentially the same (`metastoreState`). We can separate them but it wouldn't make a big difference unless we decompose the commands themselves.

> `RaftLeaderRoutine`

Agreed that we need this. We now have a third component that wants to be run only on the leader. We could benefit from a construct for facilitating this.